### DISCUSSION TITLE
How to load records from one database to another?

### DISCUSSION THREAD

**Discussion by maticjosip** [sqlalchemy/sqlalchemy#9431] [Usage Questions] [answered]
Hi everyone,

could someone tell me the proper way to load records from one database to another?

I tried to load it with the code code below, the code works until this line:
```
subquery = self.src_session.query(self.src_model).filter(self.src_model.c.Code.like('C%')).limit(1)
```
From this line I get the error: 
> [ODBC Driver 17 for SQL Server][SQL Server]Invalid object name 'SrcSchema.Usz_DRG_Diagnosis'.
```
insert_statement = insert(self.dest_model).from_select(self.dest_model.columns, subquery)
```
My intuitien tells me here should the dest_session also be referenced, but I'm not sure how it should look like.

The complete code:
```
src_map_registry = registry()
dest_map_registry = registry()


src_usz_drg_diagnosis = Table(
    "Usz_DRG_Diagnosis",
    src_map_registry.metadata,
    Column('Diagnosis_ID', NVARCHAR(400), primary_key=True),
    Column('Version', NVARCHAR(400)),
    Column('Modifier', NVARCHAR(400)),
    schema="SrcSchema",
)

dest_usz_drg_diagnosis = Table(
    "Usz_DRG_Diagnosis",
    dest_map_registry.metadata,
    Column('Diagnosis_ID', NVARCHAR(400), primary_key=True),
    Column('Version', NVARCHAR(400)),
    Column('Modifier', NVARCHAR(400)),
    schema="DestSchema",
)

@dataclass
class SrcUszDRGDiagnosis:
    Diagnosis_ID: str
    Version: str
    Modifier: str

@dataclass
class DestUszDRGDiagnosis:
    Diagnosis_ID: str
    Version: str
    Modifier: str


src_engine = create_engine("mssql+pyodbc://USR:PWD@serverName/SourceDB?Trusted_Connection=No&driver=ODBC+Driver+17+for+SQL+Server")
dest_engine = create_engine("mssql+pyodbc://USR:PWD@serverName/DestinationDB?Trusted_Connection=No&driver=ODBC+Driver+17+for+SQL+Server"))

SRC_MSSQL_SESSION_FACTORY = sessionmaker(bind=src_engine)
DEST_MSSQL_SESSION_FACTORY = sessionmaker(bind=dest_engine)


def main():
    src_session = SRC_MSSQL_SESSION_FACTORY()
    dest_session = DEST_MSSQL_SESSION_FACTORY()
    start_mappers()
    sadl = SADeltaLoad(src_engine, src_session, dest_session, src_usz_drg_diagnosis, dest_usz_drg_diagnosis)
    sadl.call()
	

def start_mappers():
    src_map_registry.map_imperatively(SrcUszDRGDiagnosis, src_usz_drg_diagnosis)
    dest_map_registry.map_imperatively(DestUszDRGDiagnosis, dest_usz_drg_diagnosis)

    src_map_registry.metadata.create_all(bind=src_engine)
    dest_map_registry.metadata.create_all(bind=dest_engine)


class SADeltaLoad:
    def __init__(self, src_engine, src_session, dest_session, src_model, dest_model):
        self.src_engine = src_engine
        self.src_session = src_session
        self.dest_session = dest_session
        self.src_model = aliased(src_model, name='src_model_alias')
        self.dest_model = dest_model

    def call(self):
	""" executes methods in correct order """
        self.insert_all_records_with_starting_c_src_to_dest()
		
    def insert_all_records_with_starting_c_src_to_dest(self):
	""" Insets records from one db to another """
        subquery = self.src_session.query(self.src_model).filter(self.src_model.c.Code.like('C%')).limit(1)
        insert_statement = insert(self.dest_model).from_select(self.dest_model.columns, subquery)
        self.dest_session.execute(insert_statement)
```

Thanks for your help!

**Comment by CaselIT** [comment, marked as answer]
Hi,

What you are doing here is running a query similar to the following
```sql
insert into dest_table select * from src_table
```

if the destination db does not see the source db, then that sql cannot work.

Did you want to load the table rows from the source db in python and then save then in the destination db or were you expecting that a single sql query executed on the destination db would do that?

**Reply by maticjosip** [reply]
Hi,
yes, the idea was to load all records with a filter condition, temporarily saved in "subquery", then insert into the destination table.
Is it necessary to put it into a list of dictionaries then insert it into the destination database, like this?

```
subquery = select(src_model).where(src_model.c.Code.like('C%')).limit(10).all()
dist_list = subquery.to_dict('records')
session.execute(insert(stg_model), dict_list)
```
The subquery would be already a list with .all(), just not a list of dictionaries.

**Reply by CaselIT** [reply]
That's not an sqlalchemy syntax, but yes the general ides if you plan on loading all records is to somewhat similar.

I'm not sure if you can directly use `result.mappings().all()` in the execute, of if you need to convert each row to a dict

Ref: https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.mappings

**Reply by maticjosip** [reply]
how would you do a load from one table to another if both tables are in different database/server? Can you pinpoint me to a tutorial?

**Reply by zzzeek** [reply]
if they are the same kind of database version I would use the database's dump/restore utilities.   e.g. https://www.postgresql.org/docs/current/app-pgdump.html

**Reply by maticjosip** [reply]
it is the same database version (MS SQL Server), but in other projects other db versions could be used.

**Reply by zzzeek** [reply]
are you creating a generic database copying utility?    Or is this a specific set of models?   What are the target backends covered?   What is the scale of data being handled (e.g. what scale of rows within tables) ?   does the tool need to generate schemas across databases or just copy data ?    

this is a very open ended problem and the most simple approach should be used based on what you actually need to accomplish.    If you just want to move data between two backends, create two engines, run a select from one into memory, then turn the rows into INSERT to send out to the other database.  if you have millions of rows, then you need to batch those rows.

**Reply by maticjosip** [reply]
first of all the code should work :)

how do you turn the rows into INSERT? 

When I execute for example:
`subquery = src_session.query(src_model).filter(src_model.c.Code.like('C%')).all()`
I get a list of tuples, how can i insert it without a for loop? e.g. execute_many/bulk_insert

I tried this:
`dest_session.execute(insert(dest_model), subquery)`
and this:
`dest_session.execute(dest_model.insert().values(subquery))`

Both do not work, is it necessary to transform the list of tuples into a list of dictionaries or is there another solution?

**Reply by zzzeek** [reply]
are you looking to use bulk insert routines or unit of work routines?

for the difference, first read for unit of work:  https://docs.sqlalchemy.org/en/20/tutorial/orm_data_manipulation.html

then read: https://docs.sqlalchemy.org/en/20/tutorial/orm_data_manipulation.html#bulk-multi-row-insert-upsert-update-and-delete

Lately there seem to be new users who are unaware of the unit of work and I'm not really sure why that is.

**Reply by zzzeek** [reply]
more generally what has been your primary source of SQLAlchemy learning so far?   is there a tutorial out there people are seeing that doesn't talk about session unit of work?

**Reply by CaselIT** [reply]
> how would you do a load from one table to another if both tables are in different database/server? Can you pinpoint me to a tutorial?

the most simple thing you can do is
```py
rows_to_insert = conn_source_db.execute(query_source_table).mappings().all()
conn_dest_db.execute(dest_table.insert(), rows_to_insert)
```

The above works for a small number of rows (let's say 10k max). If you have a lot of rows you should look into baching the rows using something like https://docs.sqlalchemy.org/en/20/core/connections.html#streaming-with-a-fixed-buffer-via-yield-per

as mike mentioned for bulk migrations proper migration tools are faster and probably less prone to errors

**Reply by maticjosip** [reply]
Hi, I'm sorry for the late reply! This week was crazy at work. Thanks for the answers, i read something about unit of work but never used it, seen it from a colleague, while i was only using pyodbc and wanted to learn something that might improve my daily business.

I managed to run the code now but only with a for loop going through each record in the subselect.
Here is the code:

```
    def insert_all_records_with_starting_c_src_to_dest(self):
        subquery = self.src_session.query(self.src_model).filter(self.src_model.c.Code.like('C%')).limit(10).all()
        for row in subquery:
            self.dest_session.execute(self.dest_model.insert().values(row.tuple()))
        self.dest_session.commit()
```

Kind of disappointed as I tried to avoid loops in my code.

it is strange when i use `self.dest_session.execute(insert(self.dest_model), subquery)` it gives me the error that _list can only contain tuples or dictionaries_. It would be really neat to bulk load subselects with the .all() functionality, the same when i bulk load dictionaries. The issue is that .all() returns a list of sqlalchemy.engine.rows, if it would return tuples or dictionaries it would be amazing.

Here a stackoverflow link with the same issue: https://stackoverflow.com/questions/1958219/how-to-convert-sqlalchemy-row-object-to-a-python-dict?page=1&tab=scoredesc#tab-top

Thanks again and wishing a great weekend!

**Reply by CaselIT** [reply]
Note that in the example above I did exactly that and it works without issues. Please try it

**Reply by maticjosip** [reply]
Here is the working code:
```
    def insert_all_records_with_starting_c_src_to_dest(self):
        subquery = self.src_session.query(self.src_model).filter(self.src_model.c.Code.like('C%')).limit(10)
        rows_to_insert = self.src_session.execute(subquery).mappings().all()
        self.dest_session.execute(self.dest_model.insert(), rows_to_insert)
        self.dest_session.commit()
```

Thanks for your time!

**Reply by CaselIT** [reply]
great, thanks for confirming!
